linq4j/src/main/java/org/apache/calcite/linq4j/ExtendedEnumerable.java

```diff
@@ -541,15 +541,47 @@
    * For each row of the current enumerable returns the correlated rows
    * from the {@code inner} enumerable (nested loops join).
    *
+   * @deprecated Use {@link #correlateJoin(JoinType, Function1, Function2)}
+   *
    * @param joinType inner, left, semi or anti join type
    * @param inner generator of inner enumerable
    * @param resultSelector selector of the result. For semi/anti join
    *                       inner argument is always null.
    */
+  @Deprecated // to be removed before 2.0
   <TInner, TResult> Enumerable<TResult> correlateJoin(
       CorrelateJoinType joinType, Function1<TSource, Enumerable<TInner>> inner,
       Function2<TSource, TInner, TResult> resultSelector);
 
+  /**
+   * For each row of the current enumerable returns the correlated rows
+   * from the {@code inner} enumerable (nested loops join).
+   *
+   * @param joinType inner, left, semi or anti join type
+   * @param inner generator of inner enumerable
+   * @param resultSelector selector of the result. For semi/anti join
+   *                       inner argument is always null.
+   */
+  @SuppressWarnings("deprecation")
+  default <TInner, TResult> Enumerable<TResult> correlateJoin(
+      JoinType joinType, Function1<TSource, Enumerable<TInner>> inner,
+      Function2<TSource, TInner, TResult> resultSelector) {
+    // temporary default implementation for backwards compatibility
```

Comment:
What are we going to do when we remove the `correlateJoin(CorrelateJoinType...` method above?

Reply:
My idea was that whenever we remove the method above, we will remove this default implementation.

Reply:
It seems the consensus is to keep CorrelateJoinType as deprecated for one minor release (1.20), and then remove it completely for 1.21. At that point, this default implementation will be removed too.